metronome/checkpointing/interpreter/src/io/iohk/metronome/checkpointing/interpreter/InterpreterService.scala

```diff
@@ -0,0 +1,161 @@
+
+          case req @ CreateBlockBodyRequest(requestId, ledger, mempool) =>
+            respondWith(
+              req,
+              interpreterRpc.createBlockBody(ledger, mempool),
+              CreateBlockBodyResponse(requestId, _)
+            )
+
+          case req @ ValidateBlockBodyRequest(requestId, blockBody, ledger) =>
+            respondWith(
+              req,
+              interpreterRpc.validateBlockBody(blockBody, ledger),
+              ValidateBlockBodyResponse(requestId, _)
+            )
+
+          case req @ NewCheckpointCertificateRequest(_, cert) =>
+            noResponse(
+              req,
+              interpreterRpc.newCheckpointCertificate(cert)
+            )
+        }
+        .completedL
+    }
+
+    private def respondWith[
+        Req <: InterpreterMessage with Request with FromService,
+        Res <: InterpreterMessage with Response with FromInterpreter,
+        A
+    ](
+        request: Req,
+        maybeResult: F[Option[A]],
+        toResponse: A => Res
+    )(implicit ev: RPCPair.Aux[Req, Res]): F[Unit] =
+      Concurrent[F].start {
+        Concurrent[F]
+          .race(
+            Timer[F].sleep(timeout),
+            maybeResult
+          )
+          .flatMap {
+            case Left(_) =>
+              tracer(InterpreterTimeout(request))
+            case Right(None) =>
+              // The response types contain non-optional values.
+              // The optional semantics on the Service side are
+              // to be achieved using an `RPCTracker`.
+              ().pure[F]
+            case Right(Some(result)) =>
+              sendMessage(toResponse(result))
+          }
+          .handleErrorWith { case NonFatal(ex) =>
+            tracer(InterpreterError(request, ex))
+          }
+      }.void
+
+    private def noResponse(
+        request: InterpreterMessage
+          with Request
+          with FromService
+          with NoResponse,
+        command: F[Unit]
+    ): F[Unit] =
+      Concurrent[F].start {
+        command.handleErrorWith { case NonFatal(ex) =>
+          tracer(InterpreterError(request, ex))
+        }
+      }.void
+
+    private def notify(
+        f: RequestId => InterpreterMessage with FromInterpreter with NoResponse
+    ): F[Unit] =
+      for {
+        requestId <- RequestId[F]
+        message = f(requestId)
+        _ <- sendMessage(message)
+      } yield ()
+
+    private def sendMessage(
+        message: InterpreterMessage with FromInterpreter
+    ): F[Unit] =
+      localConnectionManager.sendMessage(message).flatMap {
+        case Left(_)  => tracer(ServiceUnavailable(message))
+        case Right(_) => ().pure[F]
+      }
+  }
+
+  /** Start processing Service messages and in the background, delegating to the `interpreterRPC`,
+    * which is the domain specific Interpreter implementation in the host system.
+    *
+    * Returns a `ServiceRPC` instance that the host system can use to send notifications to the Service.
```

Comment:
What is the "host system"? What is the "Service"?

Reply:
By the host system I meant that this is a library imported into the PoW node. The Service, as usual, is the BFT+Checkpointing combo. 

Should we just saw "PoW node"?

Reply:
Or "host system (ie. the PoW node)" if you think the more generic term can be meaningful in another way.

Reply:
Thanks, done.